src/Database/LSMTree/Internal/WriteBuffer.hs

```diff
@@ -53,55 +62,100 @@ import           Prelude hiding (lookup)
+    -- | Keeps track of the total size of keys and values in the buffer.
+    -- This means reconstructing the 'WB' constructor on every update.
+  , writeBufferSizeBytes :: {-# UNPACK #-} !Int
+  }
+  deriving stock (Eq, Show)
+
+instance NFData WriteBuffer where
+  rnf (WB m _) = rnf m
+
+writeBufferInvariant :: WriteBuffer -> Bool
+writeBufferInvariant (WB m s) =
+    s == sum (sizeofKey <$> keys) + sum (sizeofValue <$> values)
+  where
+    keys = Map.keys m
+    values = foldMap (onValue [] pure) m :: [SerialisedValue]
 
 empty :: WriteBuffer
-empty = WB Map.empty
+empty = WB Map.empty 0
 
 numEntries :: WriteBuffer -> NumEntries
-numEntries (WB m) = NumEntries (Map.size m)
+numEntries (WB m _) = NumEntries (Map.size m)
+
+sizeInBytes :: WriteBuffer -> Int
+sizeInBytes = writeBufferSizeBytes
+
+fromMap ::
+     Map SerialisedKey (Entry SerialisedValue SerialisedBlob)
+  -> WriteBuffer
+fromMap m = WB m (Map.foldlWithKey' (\s k e -> s + sizeofKOp k e) 0 m)
+  where
+    sizeofKOp k e = sizeofKey k + onValue 0 sizeofValue e
+
+fromList ::
+     (SerialisedValue -> SerialisedValue -> SerialisedValue) -- ^ merge function
+  -> [(SerialisedKey, Entry SerialisedValue SerialisedBlob)]
+  -> WriteBuffer
+fromList f es = fromMap (Map.fromListWith (combine f) es)
 
 -- | \( O(n) \)
-content :: WriteBuffer ->
-           [(SerialisedKey, Entry SerialisedValue SerialisedBlob)]
-content (WB m) = Map.assocs m
+toList :: WriteBuffer -> [(SerialisedKey, Entry SerialisedValue SerialisedBlob)]
+toList (WB m _) = Map.assocs m
 
 {-------------------------------------------------------------------------------
   Updates
 -------------------------------------------------------------------------------}
 
+-- TODO(optimisation): make sure the core looks reasonable, e.g. not allocating
+-- 'WB' for every entry.
+addEntries ::
+     Foldable f
+  => (SerialisedValue -> SerialisedValue -> SerialisedValue) -- ^ merge function
+  -> f (SerialisedKey, Entry SerialisedValue SerialisedBlob)
+  -> WriteBuffer
+  -> WriteBuffer
+addEntries f es wb = foldl' (flip (uncurry (addEntry f))) wb (Foldable.toList es)
+
 addEntry ::
      (SerialisedValue -> SerialisedValue -> SerialisedValue) -- ^ merge function
   -> SerialisedKey
   -> Entry SerialisedValue SerialisedBlob
   -> WriteBuffer
   -> WriteBuffer
-addEntry f k e (WB wb) =
-    WB (Map.insertWith (combine f) k e wb)
+addEntry f k e (WB wb s) =
+    let (!wb', !s') = runState (insert k wb) s
+    in WB wb' s'
+  where
+    -- TODO: this seems inelegant, but we want to avoid traversing the Map twice
+    insert = Map.alterF $ (fmap Just .) $ \case
+        Nothing -> do
+          modify' (+ (sizeofKey k + sizeofEntry e))
+          return e
+        Just old -> do
+          let !new = combine f e old
+          -- don't count key (it was already there), only count value difference
+          modify' (+ (sizeofEntry new - sizeofEntry old))
+          return new
+    sizeofEntry = onValue 0 sizeofValue
```

Comment:
I think that technically also a `Writer` would be sufficient as a monad here.

Reply:
I think even that a the `(Int,)` functor should be enough

Reply:
Seems like the `State` is more performant than the `(Int,)` version, so I'll keep it as is

Reply:
The `Writer` version is slightly quicker, so I'll change it to that